share-coordinator/src/main/java/org/apache/kafka/coordinator/share/PersisterStateBatchCombiner.java

```diff
@@ -0,0 +1,401 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.coordinator.share;
+
+import org.apache.kafka.server.share.PersisterStateBatch;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.TreeSet;
+
+public class PersisterStateBatchCombiner {
+    private List<PersisterStateBatch> combinedBatchList;    // link between pruning and merging
+    private final long startOffset;
+    private TreeSet<PersisterStateBatch> sortedBatches;
+    private List<PersisterStateBatch> finalBatchList;   // final list is built here
+
+    public PersisterStateBatchCombiner(
+        List<PersisterStateBatch> batchesSoFar,
+        List<PersisterStateBatch> newBatches,
+        long startOffset
+    ) {
+        initializeCombinedList(batchesSoFar, newBatches);
+        finalBatchList = new ArrayList<>(combinedBatchList.size() * 2);   // heuristic size
```

Comment:
ArrayList grows by 50% when full, so over-allocating 100% might be a bit much. Do we have any idea how much overhead the combiner algorithm normally requires?

Maybe we can reduce this to 1.5

Reply:
merging 2 batches can result in 3 batches at max so 50% is reasonable